python/pyspark/sql/dataframe.py

```diff
@@ -3928,6 +3928,71 @@ def dropDuplicates(self, subset: Optional[List[str]] = None) -> "DataFrame":
             jdf = self._jdf.dropDuplicates(self._jseq(subset))
         return DataFrame(jdf, self.sparkSession)
 
+    def dropDuplicatesWithinWatermark(self, subset: Optional[List[str]] = None) -> "DataFrame":
+        """Return a new :class:`DataFrame` with duplicate rows removed,
+         optionally only considering certain columns, within watermark.
+
+        For a static batch :class:`DataFrame`, it just drops duplicate rows. For a streaming
+        :class:`DataFrame`, this will keep all data across triggers as intermediate state to drop
+        duplicated rows. The state will be kept to guarantee the semantic, "Events are deduplicated
+        as long as the time distance of earliest and latest events are smaller than the delay
+        threshold of watermark." The watermark for the input :class:`DataFrame` must be set via
+        :func:`withWatermark`. Users are encouraged to set the delay threshold of watermark longer
+        than max timestamp differences among duplicated events. In addition, too late data older
+        than watermark will be dropped.
+
+         .. versionadded:: 3.5.0
+
+         Parameters
+         ----------
+         subset : List of column names, optional
+             List of columns to use for duplicate comparison (default All columns).
+
+         Returns
+         -------
+         :class:`DataFrame`
+             DataFrame without duplicates.
+
+         Examples
+         --------
+         >>> from pyspark.sql import Row
+         >>> df = spark.createDataFrame([
+         ...     Row(name='Alice', age=5, height=80),
+         ...     Row(name='Alice', age=5, height=80),
+         ...     Row(name='Alice', age=10, height=80)
+         ... ])
+
+         Deduplicate the same rows.
+
+         >>> df.dropDuplicatesWithinWatermark().show()
```

Comment:
I'll look into it. I blindly followed dropDuplicates() but we'd probably want to emphasize the case of streaming for this API.

Reply:
I can't find the code example to add the code in example section but not evaluating as test. It would be tricky if we have to run streaming query from there and also have to validate something.

@HyukjinKwon Would you mind if I ask for kindly guiding about pyspark API doc? Are these codes in example section still executed as tests? If they are, is there a way to prevent it for several lines or per method? Thanks in advance!

Reply:
Updated. I just removed the example (test) for batch case since it sounds to be obvious. I can split example for batch and streaming and add back batch queries if we want to.

Reply:
Also, even if it is a batch example, better to set watermark. It will serve as good example for users, since most of them will be looking to use in streaming context.